fedimint-core/src/epoch.rs

```diff
@@ -12,14 +12,29 @@ use threshold_crypto::{PublicKey, PublicKeySet, Signature, SignatureShare};
 
 use crate::transaction::Transaction;
 
+/// All the items that may be produced during a consensus epoch
 #[derive(Debug, Clone, Eq, PartialEq, Hash, UnzipConsensus, Encodable, Decodable)]
 pub enum ConsensusItem {
+    /// Fed shutdown occurs once a threshold of peers signal (for consensus
+    /// upgrades)
+    ShutdownSignal(ShutdownReason),
```

Comment:
I was imagining that this will be more of a `ConsensusUpgrade {  versions: GlobalConsensusVersion, height: EpochHeight }` kind of thing, that will cause the `fedimintd` to quit when out of data.

I don't think we need/want a "general purpose shutdown mechanism" in the consensus. IMO it should only contain consensus-related items (which version upgrade is very much a part of).

Reply:
I could rename it, there probably aren't other consensus shutdown cases.  Where would the `GlobalConsensusVersion` come from?

Reply:
> I could rename it, there probably aren't other consensus shutdown cases. 

IMO, the way to think about it is in consensus terms. Eventually we might not even require a shutdown and fedimint might just smoothly switch to new consensus versions, etc. Shutdown is just an implementation detail of how we handle consensus upgrades currently, no?

>  Where would the GlobalConsensusVersion come from?

Hmmm... Maybe it should not be a version, but `ConsensusChange { consensus_hash: }`? With a whole consensus config hash? That would allow consensus changes without consensus upgrades, which is more general.

I'm not deep into discussion here. Eventually I guess guardian will just go to the UI, click some "propose consensus rule change", get some UI to add modules, change config values, add/remove members, etc. This will also include an epoch (date estimate). The proposal will be send to other fed members and trigger a notification to their guardians and they will see the proposal and either accept and reject? When quorum is reached the upgrade is locked and all other fedimints will follow (even if their guardians don't act / reject it).

Reply:
But currently we haven no such UI, so I'm not sure how we are planning to trigger it. Some `fedimint-cli` where one passes an epoch + new config?

Reply:
Yes, eventually we'd want it to be built into the UI.

For now, users will probably need to run a script that will backup and update any consensus config changes.  Similar to DB changes, you'd need to be able to roll-back and some migrations can be automated (also some will require running DKG).

Perhaps we should run these migrations while the fed is still running (generating new configs without using them) so that we can all confirm we have the same consensus hash?  But that would also require the new binary is available to run them while the old binary is still running.  For now it's easier just to agree to shutdown and coordinate the changes out-of-band.

Reply:
>  For now it's easier just to agree to shutdown and coordinate the changes out-of-band.

I see. OK. I guess by making it a consensus item, we can make even peers that are not around stop.

In that case I would suggest introducing a `shutdown-id`.

After performing upgrade procedure, guardian would have to restart `fedimintd` with a `--consensus-continue <shutdown-id>` or something like that to make sure operators don't get it restarted by accident. Unknown `--consensus-continue <shutdown-id>` should be ignored (with a warning) and after consensus was restarted, `--consensus-continue` should no longer be required.

The "Reason" in the shutdown seems kind of meh. If everything is out of band, I don't see why this wouldn't be.

Reply:
> After performing upgrade procedure, guardian would have to restart `fedimintd` with a `--consensus-continue <shutdown-id>` or something like that to make sure operators don't get it restarted by accident. Unknown `--consensus-continue <shutdown-id>` should be ignored (with a warning) and after consensus was restarted, `--consensus-continue` should no longer be required.

Otherwise, I don't see currently anything that would prevent `fedimintd` from just being restarted by `systemd` or other scripts.

Reply:
I second that idea, automatic systemd restarts are a concern.

Reply:
Trying to understand the process we are agreeing upon:

1. Guardians manually trigger an upgrade via an authenticated API, submitted as a consensus item
2. Once a threshold is reached a deterministic shutdown id is persisted in the DB and `fedimintd` shuts down for everyone at the same epoch
3. Guardians manually run the upgrade (config/binary changes) then pass in the shutdown id as an arg to start `fedimintd` again

Reply:
Exactly, that's how I understand it too, this will prevent accidental restarts